File: app-java/src/test/java/tech/thdev/java_udemy_sample/class_02/_03_Open.java
package tech.thdev.java_udemy_sample.class_02;

/**
 * Created by tae-hwan on 09/10/2016.
 *
 * Kotlin에서는 open을 통해 함수 재정의를 일부 허용 해준다
 *
 * java는 기본이 open이라서 final을 통해서 함수 재정의를 막을 수 있다
 *
 * 아래 예는 fianl을 통해 함수 재정의를 막는 예제이다
 */

public /*final*/ class _03_Open {

    public void methodA() {
        System.out.println("methodA");
    }

    // 함수 재정의 불가능
    public final void methodB() {
        System.out.println("methodB");
    }
}

class FinalSample extends _03_Open {

    @Override
    public void methodA() {
        System.out.println("");
    }
}

class FinalSampe2 extends FinalSample {
    @Override
    public void methodA() {
        super.methodA();
    }
}

File: app-kotlin/src/test/java/tech/thdev/kotlin_udemy_sample/class_02/_03_Open.kt
package tech.thdev.kotlin_udemy_sample.class_02

/**
 * Created by tae-hwan on 09/10/2016.
 */

// kotlin의 Abstract는 함수의 재 정의가 기본적으로 불가능
// open class를 생성하여야 재 정의 가능한 형태의 함수가 생성됩니다

open class _03_Open {

    open fun methodA() {}
    fun methodB() {
        println("final method ...")
    }
}

class OpenSample : _03_Open() {

    override fun methodA() {
        println("Functions of financial")
    }

    // methodB는 final이라서 함수 재 정의가 불가능
//    override fun methodB() {
//        // final and cannot be override
//    }
}

/**
 * OpenSample의 경우 final이기 때문에 클래스 재 정의가 불가능
 */
//class OpenSample2 : OpenSample() {
//
//}